README.md
[![Documentation Status](https://readthedocs.org/projects/smc-python/badge/?version=latest)](http://smc-python.readthedocs.io/en/latest/?badge=latest) [![GitHub version](https://badge.fury.io/gh/gabstopper%2Fsmc-python.svg)](https://badge.fury.io/gh/gabstopper%2Fsmc-python)

### smc-python

Python based library to provide the ability to interact with the Stonesoft Management Center API.
Provides automation capabilities for any environment that interact with the SMC remotely.

Some of the functionality you get with the SMC Python API:

* Create any engine types; single firewall, cluster firewalls, ips engines, layer 2 firewalls, master engine and virtual engines.
* Engine operations such as enabling/disabling AV, GTI, default NAT, Contact Addresses, etc
* Interface configurations
* Routing configurations (OSPF, BGP, Static, Antispoofing)
* Engine level commands such as rebooting, going offline, policy push, enable/disable SSH, etc.
* Create and modify all network element objects such as Host, Network, Address Ranges, Domain Names, etc.
* Policy control (create rules, delete rules) for layer 3 firewall policies
* VPN Policy control and creation
* Management / Log Server settings configuration
* Admin User creation and modification
* System level controls; update system packages, update engines, global blacklisting, etc
* Tasks
* Search operations for any object type by name, href and by filter
* Collections interface to view all objects by type


##### Requirements

Python 2.7

Python 3.4, 3.5

Requests >= 2.12.0

Security Management Center version 6.0, 6.1, 6.1.1, 6.1.2, 6.2, 6.2.1

##### Getting Started

Installing package

Use pip:

`pip install git+https://github.com/gabstopper/smc-python.git`

Specific version:

`pip install -e git://github.com/gabstopper/smc-python.git@v0.4.2#egg=smc-python-0.4.2`

Download the latest tarball: [smc-python](https://github.com/gabstopper/smc-python/archive/master.zip), unzip and run:

`python setup.py install`

##### Testing

Included are a variety of test example scripts that leverage the API to do various tasks in /examples

##### Basics

Before any commands are run, you must obtain a login session. Once commands are complete, call smc.logout() to remove the active session. To obtain the api_key, log in to the Stonesoft Management Center and create an API client with the proper privileges.

```python
from smc import session

session.login(url='http://1.1.1.1:8082', api_key='xxxxxxxxxxxxx')
....do stuff....
session.logout()
```

Or log in to a specific Admin Domain and use a specific version of the API:

```python
session.login(url='http://1.1.1.1:8082', api_key='xxxxxxxxxxxxx',
			  domain=mydomain, api_version=6.2)
....do stuff....
session.logout()
```

Once a valid session is obtained, it will be re-used for each operation for the duration of the sessions validity, or until the program is exited.

 
Please see the read-the-docs documentation above for a full explanation and technical reference on available API classes.

[View Documentation on Read The Docs](http://smc-python.readthedocs.io/en/latest/?badge=latest)

HISTORY.rst
Release History
===============

0.5.6
-----

.. note:: For older release release history information, see CHANGELOG. All future documentation will be logged
	in this document.

**Improvements**

- Support SMC 6.3:
    - Support for L2 interface policies (Inline L2, IPS and Capture interfaces on L3 engine)
    - Route based VPN support, IPSEC wrapped RBVPN and GRE Tunnel/Transport/No Encryption VPN.
- SMC 6.3 API only supports TLSv1.2 or greater, ensure your openssl version supports TLSv1.2. This can be done
  by: openssl s_client -connect <smc_ip>:8082 -tls1_2
- Simplified generic Search (`smc.base.collections.Search`) to be uniform with ElementCollection.
- Simplify API reference documentation
- SMC login using environment variables. See session documentation for more info.
- Rule counters on all Policy types
- Proxy or static type required when adding arp entry to interface
- Add simple .get() method on Element. This simplifies determining if the element by name exists. For example,
  Host.get('kali') would raise ElementNotFound if it doesn't exist. Prior to this, you would have to search
  for the element and attempt to access and element resource before receiving the ElementNotFound message,
  i.e. host = Host('kali'); host.address. The 'get()' method still returns an 'un-inflated' instance (only meta
  data).
- Deprecation warnings are now generated for functions in `smc.core.engine.interfaces`:
  `add_single_node_interface`, `add_node_interface`, `add_vlan_to_node_interface`, `add_ipaddress_to_vlan_interface`.
  These functions will eventually be deprecated. As of version 6.3, SMC engines can now support both layer 2 and
  layer 3 interfaces on the same engine. New interface functions added: `add_layer3_vlan_interface`, `add_layer3_interface`,
  `add_inline_ips_interface`, `add_inline_l2fw_interface`.
- New element types: URLCategory, URLCategoryGroup, ICMPServiceGroup


.. important:: Renamed `smc.vpn.policy.VPNPolicy` to `smc.vpn.policy.PolicyVPN`

**Bugfixes**

- HTTP GET was treating a 204 response as an error, fix to treat No Content response as success.
- Fix help() on dynamic `create_collection` class so constructor methods are proxied properly
- Raise SMCConnectionError when non-HTTP 200 error code presented from SMC when retrieving entry points
- Sending empty payload on POST request with parameters might cause validation error. Do not submit empty
  dict with POST requests.
  
0.5.8
-----

**Improvements**

- Support for SMC version 6.3.0, 6.3.1 and 6.3.2
- Add ``case_sensitive`` key word to filtered queries. This requires SMC 6.3+. Set this as a kwarg when making
  the query: Host.objects.filter('myhost', case_sensitive=False). Default: case_sensitive=True.
- Optimize retrieval of nodes by serializing engine node data versus making a call to the engine links. This eliminates
  the query to get the node links and a query for each node that needs to be operated on, or node payload required.
- Add ApplianceInfo and link on node to retrieve appliance related info
- GatewayTunnel implemented on PolicyVPN for setting preshared key, enabling/disabling specific tunnel endpoints
- BGP node added to engine. Add full create/modify/delete capability by reference: engine.bgp.is_enabled, etc. Added to
  provide modular configuration to BGP.
- OSPF node added to engine. Add full create/modify/delete capability by reference: engine.ospf.is_enabled, etc.
- merging lists on element update will now filter out duplicate entries before potentially updating. The SMC API protects
  against this but validation moved into element update function saving potential exception on PUT
- get_or_create and update_or_create return classmethod get for elements that are considered read-only; i.e. do not have
  a `create` classmethod.
- update_or_create will now check the provided key/value pairs before updating the specified element. This is to make
  the modification more idempotent. If the retrieved element exists and has the same value (based on current ETag), then
  do not modify.
- Optimization of resolved alias retrieval from the engine. Instead of retrieving all aliases and resolving the alias
  reference, first retrieve the entire list of aliases (1 query) and then correlate to resolved alias references. This
  amounts to reducing the number of queries to retrieve a single engines aliases from ~60 to 3.
- set_stream_logger and set_file_logger attached to smc.api.session.Session() as convenience functions.
- Optimize logging at request level, more clear output
- Simplify interface creating where zone or logical interface is needed. Now zone/logical interfaces can be provided
  as either name (if they don't exist, they will be created), as href, or as Zone/LogicalInterface instances.
- New engine level resources: antivirus, file_reputation, sidewinder_proxy, sandbox and url_filtering, policy_routing,
  dns and default nat added as engine resources. Previous functions nested in smc.core.properties.AddOns set to deprecated
  and will be removed in the near future.
- Added support for adding DNS Server entries to engines based on elements (previously only IP addresses were supported).
- TLS Server Credentials supported for inbound SSL decryption, add to engine from engine.tls_inspection resource.
- Add create_hook to ElementCreator to intercept json before submitting to SMC server. See smc.base.decorators.create_hook
  for more info.
- Added engine.interface_options node for settings related to setting primary mgt, backup mgt, primary hearbeat, and backup heartbeat
  rather than having them nested on the PhysicalInterface. These can be called directly from the engine which removes ambiguity in how
  these settings are modified. Previous versions they could be called directly (i.e. engine.physical_interface.set_primary_mgt() however
  required unnecessary plumbing. This more closely models the SMC UI configuration.

smc/search.py
'''
Created on May 1, 2016

@author: davidlepage
'''

import logging
import smc

logger = logging.getLogger(__name__)

def get_element(name, obj_type=None, use_name_field=True):
    """ Get specified element/s by name
        When specifying just name arg, all elements are searched
        Query will look like: http://1.1.1.1/elements?filter=name
        Use obj_type if you know the type of element to filter further
            Arg:
                * name: what element to search for
                * obj_type (optional): search filter by type of obj (host,router,network,etc); default=None
                * use_name_field (optional): verifies name field match in return records; default=True
            Returns:
                Json record/s representing top level element match
                None; no results
    """   
    entry_href = None
    if obj_type:
        entry_href = smc.web_api.get_entry_href(obj_type)   #specific obj type entry
        if not entry_href:
            logger.error("Object type entry point specified was not found: %s" % obj_type)
            return
    else:
        entry_href = smc.web_api.get_entry_href('elements') #entry point for all elements
    
    logger.debug("Searching for element: %s" % name)
    result = smc.web_api.http_get(entry_href + '?filter=' + name) #execute search
   
    if not result.msg: #no results returned
        logger.info("No results found for element name: %s" % name)
    else:
        if not use_name_field: # Return any host objects using name as filter (wildcard match)
            logger.debug("Returning all elements matching: %s" % name)
            return result.msg  # Return anything that was found
        else: # Only return object where name matches host object name field
            match = None #TODO: This could return multiple entries!!!
            for host in result.msg:
                if host['name'] == name:
                    match = host
                    break   #exits on first match
            if not match:
                logger.debug("No search results found for name: %s" % name)
            else:
                logger.debug("Search found direct match for element: %s, %s" % (name,match))
            return match

def get_element_by_href(name):
    """ Get specified element from fully qualified href
        For example, name could be: http://1.1.1.1:8082/elements/single_fw/119
        Args: 
            * name: href (typically reference from smc) for object to retrieve
        Returns:
            Json record representing match
    """
    result = smc.web_api.http_get(name)
    if not result.msg:
        logger.info("No results found for href, maybe a bad reference: %s" % name)
    return result.msg

def get_element_by_entry_point(name):
    """ Get specified element based on the entry point verb from SMC api
        To get the entry points available, you can call web_api.get_all_entry_points()
        For example: smc.get_element_by_entry_point('log_server')
        Args:
            * name: top level entry point name
        Returns:
            Json representation of name match
    """
    entry = smc.web_api.get_entry_href(name)
    if entry: #in case an invalid entry point is specified
        result = smc.web_api.http_get(entry)
        return result.msg #TODO: return at bottom
    else:
        logger.error("Entry point specified was not found: %s" % name)
     
if __name__ == '__main__':
    smc.web_api.login('http://172.18.1.150:8082', 'EiGpKD4QxlLJ25dbBEp20001')
    
    print "get_element-allmatchingbyname: %s" % get_element('ami')                #get all matching elements with name=ami
    print "get_element-routerobj: %s" % get_element('ami', 'router')        #get only matching host elements matching name=ami
    print "get_element-hostandallmatching: %s" % get_element('ami', 'host', False) #get all matching host elements (wildcard)
     
    '''
    print "Filter by element: %s" % filter_by_element('ami')
    print "Filter by element, not found: %s" % filter_by_element('a11124fdgdmi') 
   
    print "Filter by type: %s" % filter_by_type('single_fw', 'api-fw') 
    print "Filter by type, not found: %s" % filter_by_type('single_fw', 'blah') 
    
    print "Get Element by entry point: %s" % get_element_by_entry_point('log_server')
    print "Get Element by entry point: %s" % get_element_by_entry_point('blah')
    
    print "Get Element by href: %s" % get_element_by_href('http://172.18.1.15:8082/blah') #FIX
    print "Get Element by href: %s" % get_element_by_href('http://172.18.1.150:8082/elements/single_fw/119') #FIX
    f = smc.filter_by_type('single_fw', 'api-fw2')
    if f:
        print "filter_by_type, not found: %s" % smc.get_element_by_href(f['href'])
    '''
    
    '''route = [{ 
                     "invalid": False,
                     "ip": "172.18.1.80",
                     "key": 766,
                     "level": "gateway",
                     "link": [{"href": "http://172.18.1.150:8082/6.0/elements/single_fw/1164/routing/766",
                                "method": "GET",
                                "rel": "self",
                                "type": "routing"}],
                     "name": "172.18.1.80",
                     "read_only": False,
                     "routing_node": [{"href": "http://172.18.1.150:8082/6.0/elements/network/143",
                                        "invalid": False,
                                        "ip": "192.168.1.0/24",
                                        "key": 767,
                                        "level": "any",
                                        "link": [{"href": "http://172.18.1.150:8082/6.0/elements/single_fw/1164/routing/767",
                                                   "method": "GET",
                                                   "rel": "self",
                                                   "type": "routing"}],
                                        "name": "net-192.168.1.0/24",
                                        "read_only": False,
                                        "routing_node": [],
                                        "system": False}],
                     "system": False
        }]

    #Get FW link
    fw = smc.filter_by_type('single_fw', 'test-run') 
    print "Getting fw at: %s" % fw['href']   
    #Get FW detaild
    fw_details = smc.get_element_by_href(fw['href'])
    #pprint(fw_details['link'])
    #print any(d['rel'] == 'routing' for d in fw_details['link'])
    #Get link to routing
    link_to_routing = next(item for item in fw_details['link'] if item['rel'] == 'routing')
    print "Link to routing: %s" % link_to_routing['href']
    routing_details = smc.get_element_by_href(link_to_routing['href'])
    #print "Routing details:"
    #pprint(routing_details)
    #Find routing node in hash
    print "I'm here"
    pprint(routing_details['routing_node'][0]['routing_node'])
    print "Routing"
    a = routing_details['routing_node'][0]['routing_node'][0]
    
    b = { 'invalid': False,
        'key': 765,
        'name': 'network-3.3.3.0/24',
        'routing_node': [] }
    #pprint(a)
    print "......"
    smc.http_get('http://172.18.1.150:8082/6.0/elements/single_fw/1326/physical_interface/321')
    #Just try and PUT it anyways....
    try:
        r = smc.web_api.http_put('http://172.18.1.150:8082/6.0/elements/single_fw/1326/physical_interface/321', b, "MzIxMzk5MQ==")
    except SMCOperationFailure, e:
        print e.msg
    
    #pprint(smc.get_element_by_href('http://172.18.1.150:8082/6.0/elements/network/677'))
    #Generic routing node level
    #pprint(smc.web_api.get_all_entry_points())'''
    smc.web_api.logout()